sockeye/data_io.py

```diff
@@ -1256,23 +1278,38 @@ def fill_up(self,
             bucket_label = self.label[bucket_idx]
             num_samples = bucket_source.shape[0]
 
+            # Fill up the last batch according to the fill up policy.
+            # 'replicate' randomly samples to populate the remaining spots.
+            # 'repeate_last' instead repeats the last element.
             if num_samples % bucket_batch_size != 0:
+                rest = bucket_batch_size - num_samples % bucket_batch_size
                 if fill_up == 'replicate':
-                    rest = bucket_batch_size - num_samples % bucket_batch_size
-                    logger.info("Replicating %d random samples from %d samples in bucket %s "
-                                "to size it to multiple of %d",
-                                rest, num_samples, bucket, bucket_batch_size)
-                    random_indices_np = rs.randint(num_samples, size=rest)
-                    random_indices = mx.nd.array(random_indices_np)
-                    if isinstance(source[bucket_idx], np.ndarray):
-                        source[bucket_idx] = np.concatenate((bucket_source, bucket_source.take(random_indices_np)), axis=0)
-                    else:
-                        source[bucket_idx] = mx.nd.concat(bucket_source, bucket_source.take(random_indices), dim=0)
-                    target[bucket_idx] = mx.nd.concat(bucket_target, bucket_target.take(random_indices), dim=0)
-                    label[bucket_idx] = mx.nd.concat(bucket_label, bucket_label.take(random_indices), dim=0)
+                    logger.info("Filling bucket %s from size %d to %d by sampling with replacement",
+                                bucket, num_samples, bucket_batch_size)
+                    desired_indices_np = rs.randint(num_samples, size=rest)
+                    desired_indices = mx.nd.array(desired_indices_np)
+
+                elif fill_up == 'zeros':
+                    logger.info("Filling bucket %s from size %d to %d by repeating the last element %d %s",
+                                bucket, num_samples, bucket_batch_size, rest, inflect('time', rest))
+                    desired_indices_np = np.array([num_samples-1] * rest)
+                    desired_indices = mx.nd.array(desired_indices_np)
+
                 else:
                     raise NotImplementedError('Unknown fill-up strategy')
 
+                if isinstance(source[bucket_idx], np.ndarray):
+                    source[bucket_idx] = np.concatenate((bucket_source, bucket_source.take(desired_indices_np)), axis=0)
+                else:
+                    source[bucket_idx] = mx.nd.concat(bucket_source, bucket_source.take(desired_indices), dim=0)
+                target[bucket_idx] = mx.nd.concat(bucket_target, bucket_target.take(desired_indices), dim=0)
+                label[bucket_idx] = mx.nd.concat(bucket_label, bucket_label.take(desired_indices), dim=0)
+
+                if fill_up == 'zeros':
+                    source[bucket_idx][num_samples:,:,:] = 0
```

Comment:
might make sense to fill these with C.PAD_ID (which of course is 0).

But could you make sure that this is actually necessary? I might miss some context, but for example `RawParallelDatasetLoader.load()` creates source/target/label arrays that are filled with C.PAD_ID (0) already.

Reply:
It does seem necessary in that removing this code results in an underfilled batch that is then ignored, but I don't quite understand why. I'll look into it.

Reply:
Okay, this is clear now. `RawParallelDatasetLoader.load()` creates a list of ndarrays of shape `(num_samples, len)` for the buckets. `num_samples` is different for each bucket, obviously, and is not guaranteed to be a multiple of the batch size. That is fixed when `ParallelDataset.fill_up()` is called.

Put more succinctly, `load()` pads columns, `fill_up()` pads rows.